src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
+        Set<String> result = new HashSet<>();
+        for (BranchSource s: sources) {
+            result.add(s.getSource().getId());
+        }
+        return result;
+    }
+
+    private boolean equalButForId(SCMSource a, SCMSource b) {
+        if (!a.getClass().equals(b.getClass())) {
+            return false;
+        }
+        boolean fallBack = false;
+        Class<?> clazz = a.getClass();
+        REFLECTION:
+        while (clazz != null && SCMSource.class.isAssignableFrom(clazz) && clazz != SCMSource.class) {
+            for (Field field : clazz.getDeclaredFields()) {
+                int modifiers = field.getModifiers();
+                if (Modifier.isStatic(modifiers)) {
+                    continue;
+                }
+                if (Modifier.isTransient(modifiers)) {
+                    continue;
+                }
+                // TODO Will Java 9 modules lock setAccessible permissions? Likely would break XStream too!
+                if (!field.isAccessible()) {
+                    field.setAccessible(true);
+                }
+                if (!field.isAccessible()) {
+                    // err on the side of caution
+                    fallBack = true;
+                    break REFLECTION;
+                }
+                Object aField;
+                Object bField;
+                try {
+                    aField = field.get(a);
+                    bField = field.get(b);
+                } catch (IllegalAccessException e) {
+                    // err on the side of caution
+                    fallBack = true;
+                    break REFLECTION;
+                }
+                if (aField == null && bField == null) {
+                    // next field
+                    continue;
+                }
+                if (aField == null || bField == null) {
+                    // we know they are different
+                    return false;
+                }
+                if (aField.equals(bField)) {
+                    // next field
+                    continue;
+                }
+                if (aField instanceof Boolean || aField instanceof Number || aField instanceof String) {
+                    // no need to compare primatives by XML, we know they are different
+                    return false;
+                }
+                if (SOURCE_ID_OMITTED_XSTREAM.toXML(aField).equals(SOURCE_ID_OMITTED_XSTREAM.toXML(bField))) {
+                    // next field
+                    continue;
+                }
+                // different enough
+                return false;
+            }
+
+            clazz = clazz.getSuperclass();
+        }
+        return !fallBack || SOURCE_ID_OMITTED_XSTREAM.toXML(a).equals(SOURCE_ID_OMITTED_XSTREAM.toXML(b));
+    }
+
+    private final static XStream2 SOURCE_ID_OMITTED_XSTREAM = new XStream2();
+
+    static {
+        SOURCE_ID_OMITTED_XSTREAM.omitField(SCMSource.class, "id");
```

Comment:
Reflection code was fun to write... ok no it wasn't but it was less scary that the evil reflex stuff I was doing before I found omitField